File: src/makeRhsf.cpp
#include <string>
#include <cstring>
#include <vector>
#include <span>
#include <fstream>
#include <map>
#include <algorithm>
#include <charconv>

#include <iostream>

#include <nvector/nvector_serial.h>

#include "../include/constants.hpp"
#include "../include/readcsv.hpp"
#include "../include/Rhsf.hpp"

namespace rhsfBuilder{
    
    std::vector<term> terms; //反応項リスト
    std::map<std::string, int> termIndex; //反応速度定数名からindexへのマップ
    std::array<double, config::constantSize> rateConstants; //反応速度定数配列
    std::array<double, config::species+1> speciesData; //初期種量配列


    auto term::operator<=>(const term& other) const {
        if(reactant2 == other.reactant2){
            if(reactant1 == other.reactant1){
                return add_to <=> other.add_to;
            }
            return reactant1 <=> other.reactant1;
        }
        return reactant2 <=> other.reactant2;
    }

    std::string strip(std::string s){
        size_t start = s.find_first_not_of(" \t\n\r");
        size_t end = s.find_last_not_of(" \t\n\r");
        return (start == std::string::npos) ? "" : s.substr(start, end - start + 1);
    }

    bool is_stoiable(std::string s){
        s = strip(s);

        long v;
        auto res = std::from_chars(s.data(), s.data()+s.size(), v);
        return res.ec == std::errc() && res.ptr == s.data()+s.size();
    }
    void buildRhsf(){
        speciesData[config::species]=1.0; //ダミー種の初期量は1.0に設定
        std::vector<std::vector<std::string>> csv_data = read_csv(std::string(config::reactNetworkFile));
        std::map<std::tuple<int, int, int, int>,int> ODE; //key:(add_to, init,entering, kind), value: duplicacy
        terms.resize(0);
        auto addTerm = [&](int add_to, int init, int entering, int kind, int duplicate_count){
            std::tuple<int, int, int, int> key = std::make_tuple(add_to, init, entering, kind);
            if (ODE.contains(key)) ODE[key] += duplicate_count;
            else ODE[key] = duplicate_count; 
        };
        std::vector<std::map<std::string, std::string>> DataDict;
        for (size_t i = 1; i < csv_data.size(); ++i) {
            std::map<std::string, std::string> rowDict;
            for (size_t j = 0; j < std::min(csv_data[i].size(), csv_data[0].size()); ++j) {
                rowDict[strip(csv_data[0][j])] = strip(csv_data[i][j]);
            }
            DataDict.push_back(rowDict);
        }
        for(const auto& mp : DataDict){
            std::string kind=mp.at("kind");
            termIndex[kind] = 0;
        }
        int idx=0;
        for(auto& [key, val] : termIndex){
            val=idx;
            idx++;
        }
        for (const auto& mp : DataDict) {
            int init = std::stoi(mp.at("init_assem_id"));
            int entering = is_stoiable(mp.at("entering_assem_id")) ? std::stoi(mp.at("entering_assem_id")) : config::species;
            int product = std::stoi(mp.at("product_assem_id"));
            int leaving = is_stoiable(mp.at("leaving_assem_id")) ? std::stoi(mp.at("leaving_assem_id")) : config::species;
            int kind = termIndex.at(mp.at("kind"));
            int duplicacy = std::stoi(mp.at("duplicate_count"));
            
            addTerm(init, init, entering, kind, -duplicacy);
            if (entering != config::species) {
                addTerm(entering, init, entering, kind, -duplicacy);
            }
            addTerm(product, init, entering, kind, duplicacy);
            if (leaving != config::species) {
                addTerm(leaving, init, entering, kind, duplicacy);
            }
        }
        for (const auto& [key, duplicacy] : ODE) {
            int add_to = std::get<0>(key);
            int init = std::get<1>(key);
            int entering = std::get<2>(key);
            int kind = std::get<3>(key);
            rhsfBuilder::term t;
            t.add_to = add_to;
            t.duplicacy = duplicacy;
            t.reactant1 = init;
            t.reactant2 = entering;
            t.rateConstant = kind;
            rhsfBuilder::terms.push_back(t);
        }
        std::sort(rhsfBuilder::terms.begin(), rhsfBuilder::terms.end());
        
    }
    int rhsf(sunrealtype t, N_Vector y, N_Vector ydot, void *user_data) {
        sunrealtype* sp_ptr = N_VGetArrayPointer(y);
        auto ydotData = N_VGetArrayPointer(ydot);
        // コピー: y -> speciesData (memcpy)
        std::memcpy(rhsfBuilder::speciesData.data(), sp_ptr, config::species * sizeof(double));
        // コピー: user_data(k) -> rateConstants (memcpy)
        std::memcpy(rhsfBuilder::rateConstants.data(), user_data, config::constantSize * sizeof(double));
        std::fill(ydotData, ydotData + config::species, 0.0);
        std::span<double> ydotspan(ydotData, config::species);
        for (const auto& term : rhsfBuilder::terms) {
            term.calculate(ydotspan);
        }

        return 0;
    }
}

File: script/makeRhsf.py
import numpy as np
import pandas as pd
import csv
import pprint
from pathlib import Path

def make_rhsf(input_path: Path | None = None) -> None:
    base = Path(__file__).parent.parent
    inputFile = Path(input_path) if input_path else base / 'data' / 'classified_reactions_str.csv'
    outputFile = base / 'include' / 'Rhsf.hpp'
    df = pd.read_csv(inputFile, index_col=0)

    records = df.to_dict(orient="records")
    reaction_kinds = list(df["kind"])
    reaction_kinds = [rk.strip() if isinstance(rk, str) else rk for rk in reaction_kinds]
    reaction_kinds = list(dict.fromkeys(reaction_kinds).keys())
    reaction_kinds.sort()

    typeToIndex = {reaction_kinds[i]: i for i in range(len(reaction_kinds))}

    species = 0
    for rec in records:
        species = max(species, rec["init_assem_id"])
        species = max(species, rec["entering_assem_id"])
        species = max(species, rec["product_assem_id"])
        species = max(species, rec["leaving_assem_id"])
    species = round(species) + 1

    ODE = [dict() for _ in range(species)]

    def add_dict(mp: dict, key, value: float):
        mp[key] = mp.get(key, 0) + value

    def to_formula(tup, value):
        if tup[1] is None:
            return f"({value} * k[{typeToIndex[tup[2]]}] * sp[{tup[0]}])"
        else:
            return f"({value} * k[{typeToIndex[tup[2]]}] * sp[{tup[0]}] * sp[{tup[1]}])"

    for rec in records:
        init = rec["init_assem_id"]
        entering = None if pd.isna(rec["entering_assem_id"]) else round(rec["entering_assem_id"])
        product = rec["product_assem_id"]
        leaving = None if pd.isna(rec["leaving_assem_id"]) else round(rec["leaving_assem_id"])

        key = (init, entering, rec["kind"])
        add_dict(ODE[init], key, -rec["duplicate_count"])
        if entering is not None:
            add_dict(ODE[entering], key, -rec["duplicate_count"])
        add_dict(ODE[product], key, rec["duplicate_count"])
        if leaving is not None:
            add_dict(ODE[leaving], key, rec["duplicate_count"])

    with open(outputFile, "w") as f:
        f.write(
            "#if USE_PREGENERATED_RHSF\n\n"
            "#pragma once\n"
            "#include <array>\n"
            "#include <nvector/nvector_serial.h>\n"
            "#include \"constants.hpp\"\n\n"
            "int rhsf(sunrealtype t, N_Vector y, N_Vector ydot, void *user_data) {\n"
            "\tauto sp = N_VGetArrayPointer(y);\n"
            "\tauto ydotData = N_VGetArrayPointer(ydot);\n"
            "\tstd::array<double, config::constantSize> &k = *static_cast<std::array<double, config::constantSize>*>(user_data);\n\t"
        )
        for idx, t in enumerate(ODE):
            f.write(f"ydotData[{idx}] = ")
            first = True
            for key, val in t.items():
                if not first:
                    f.write(" + ")
                first = False
                f.write(to_formula(key, val))
            f.write(";\n\t")
        f.write("return 0;\n}\n\n#endif // USE_PREGENERATED_RHSF\n")


if __name__ == "__main__":
    make_rhsf()
